core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js

```diff
@@ -0,0 +1,168 @@
+oppia.factory('NumberWithUnitsObjectFactory', [
+  'UnitsObjectFactory', 'FractionObjectFactory',
+  function(UnitsObjectFactory, FractionObjectFactory) {
+    var NumberWithUnits = function(type, real, Fraction, Units) {
+      this.type = type;
+      this.real = real;
+      this.Fraction = Fraction;
+      this.Units = Units;
+    };
+
+    NumberWithUnits.fromRawInputString = function(rawInput) {
+      rawInput = rawInput.trim();
+      var type = '';
+      var real = '';
+      var Fraction = '';
+      var units = '';
+      var value = 0;
+      var Units = '';
+
+      // Start with digit when there is no currency unit.
+      if (rawInput.match(/^\d/)) {
+        var ind = rawInput.indexOf(' ');
+        if (ind === -1) {
+          // There is value with no units.
+          value = rawInput;
+        } else {
+          value = rawInput.substr(0, ind);
+          units = rawInput.substr(ind + 1).trim();
+        }
+      } else {
+        var ind = rawInput.indexOf(' ');
+        units = rawInput.substr(0, ind) + ' ';    // Currency unit
+        var ind2 = rawInput.indexOf(' ', ind + 1);
+        if (ind2 !== -1) {
+          value = rawInput.substr(ind + 1, ind2 - ind - 1).trim();
+          units += rawInput.substr(ind2 + 1).trim();
+        } else {
+          value = rawInput.substr(ind + 1).trim();
+          units = units.trim();
+        }
+      }
+
+      if (value.includes('/')) {
+        type = 'fraction';
+        Fraction = FractionObjectFactory.fromRawInputString(value);
+      } else {
+        type = 'real';
+        real = parseFloat(value);
+      }
+      if (units !== '') {
+        Units = UnitsObjectFactory.fromRawInputString(units);
+      }
+      return new NumberWithUnits(type, real, Fraction, Units);
+    };
+
+    return NumberWithUnits;
+  }
+]);
+
+oppia.factory('UnitsObjectFactory', [function() {
+  var Units = function(units) {
+    this.units = units;
+  };
+
+  var isunit = function(unit) {
+    return !('/*() '.includes(unit));
+  };
+
+  var stringToLexical = function(units) {
+    units += '#';
+    var unitList = [];
+    var unit = '';
+    for (var i = 0; i < units.length; i++) {
+      if ('*/()# '.includes(units[i])) {
+        if (unit.length > 0) {
+          if ((unitList.length > 0) && isunit(unitList.slice(-1).pop())) {
```

Comment:
Why do slice and pop? Why not simply check last element?

Reply:
slicing and popping does the same work of checking the last element in the list. Here, the '-1'  index to access last element is not allowed.